File: MCOINS - Coins Game.cpp
#include <bits/stdc++.h>

using namespace std;

bool dp[1000010];

void solve(int k,int l)
{
    int y = min(k,l);
    int z = max(k,l);

    for (int i=0;i<1000010;i++)
    {
        if (i < 1)
            dp[i] = false;
        else if (i < y)
            dp[i] = !dp[i-1];
        else if (i < z)
            dp[i] = !(dp[i-1] && dp[i-y]);
        else
            dp[i] = !(dp[i-1] && dp[i-y] && dp[i-z]);
    }
}

int main()
{
    int k,l,n,x;
    cin >> k >> l >> n;

    solve(k,l);

    for (int i=0;i<n;i++)
    {
        cin >> x;
        if(dp[x])
            cout << 'A';
        else
            cout << 'B';
    }
    cout << endl;
}

File: MCOINS - Coins Game.py
dp = {}

def solve(k,l):
    x = 1
    y = min(k,l)
    z = max(k,l)

    for i in range(0,1000010):
        if i < 1:
            dp[i] = False
        elif i < y:
            dp[i] = not dp[i-1]
        elif i < z:
            dp[i] = not (dp[i-x] and dp[i-y])
        else:
            dp[i] = not (dp[i-x] and dp[i-y] and dp[i-z])

k,l,n = map(int,input().split())

a = list(map(int,input().split()))

solve(k,l)

s = ""

for i in a:
    if dp[i]:
        s += 'A'
    else:
        s += 'B'

print(s)

